### PR TITLE
DONT MERGE BEFORE REVIEW: Disconnect config/state/app

### PR DESCRIPTION
**Pull request by Xiphe**
<a href="https://teamsgame.agilepainrelief.com/preview/disconnect-config">published to gh-pages</a><hr />

This is a huge refactoring targeting towards https://github.com/mlevison/high-performance-teams-game/issues/103

I'm stopping here today, will write up what changed and the ideas behind that tomorrow.

### CODE REVIEW

**Comment by mlevison** [comment]
Ok alot of this I can follow easily. 
- I can see most of the separation between UI and state -> example instead of using constants directly, accessing props that then use the constants
- I understand the elimination of GremlinId and therefore GremlinList is what declares the constants
- I understand the same changes with rounds
- I'm a little sad that losing the GremlinId, GameActionId weaken the limited typesafety in methods like nextRound and selectAction. _I assume that will also limit intellisense_
 
- I mourn the passing of gameConstants - _This change feels very odd. Anychance we can create a new version?_

I don't understand import/export type and how they differ from normal.

Hoping that makes the explanation process go faster

**Comment by Xiphe** [comment]
Cool Mark, thanks for already taking a look!

> I can see most of the separation between UI and state -> example instead of using constants directly, accessing props that then use the constants

Thats right, also the config is further separated. Before ui and state directly accessed values from the config. Now the config is passed as parameter into UI and state.

These are the main benefits we gain:

 - Game state can be executed without needing react or the UI (headless)
 - Configs can be changed easily (https://github.com/mlevison/high-performance-teams-game/issues/103#issuecomment-790055608) (This will also make the test setup much easier, you'll see)

> I understand the elimination of GremlinId and therefore GremlinList is what declares the constants
> I understand the same changes with rounds
> I'm a little sad that losing the GremlinId, GameActionId weaken the limited typesafety in methods like nextRound and selectAction. I assume that will also limit intellisense

The `GremlinId` and `ActionIds` were declared this way to enable type-checking. Since the state machine can now work with different configs, it can not have an internal understanding of the string literals otherwise it would be coupled again to the config.
I noted that the UI and state never used these IDs as literals (there is no `dispatch({ type: 'SELECT_ACTION', payload: 'SOME_ACTION_ID' })`)
The intellisense was only used in tests (and potential future programatic used) where we bypass the UI to select a specific action or gremlin.

That said, I have Ideas how to revive the types in these scenarios, let's see if they work out.

> I mourn the passing of gameConstants - This change feels very odd. Anychance we can create a new version?

Actually I'm really happy they're gone. This was the prior way of "configuring" the app and a selection of stuff I had not found a better place for :).
All the infos are now derived from the config and I think that's elegant.

```ts
export const START_CAPACITY = 10; // moved to config.round[0].effect
export const START_USER_STORY_CHANCE = 30; // moved to config.round[0].#effect
export const START_GREMLIN_CHANCE = 0; // moved to config.round[0].#effect
export const GAME_ROUNDS = 8; // derived from config.rounds.length
export const TRAILING_ROUNDS = 6;  // moved to config.trailingRounds
export const EFFECT_HIDDEN = Symbol('EFFECT_HIDDEN');  // removed 

export const TOTAL_ROUNDS = GAME_ROUNDS + TRAILING_ROUNDS; // derived from config.rounds.length + config.trailingRounds
```

That said, we could re-introduce as `config/constants.ts` but I personally don't see added value from that

```ts
export const START_CAPACITY = 10; // then use in config.round[0].effect
export const START_USER_STORY_CHANCE = 30; // then use in config.round[0].#effect
export const START_GREMLIN_CHANCE = 0; // then use in config.round[0].#effect
export const GAME_ROUNDS = 8; // then use to fill config.rounds with empty rounds if less are set
export const TRAILING_ROUNDS = 6;  // then use in config.trailingRounds
```

> I don't understand import/export type and how they differ from normal.

`import type` will not execute the module after compilation but only get the type information for the typescript compiler
it also prevents us from accidentally importing actual implementation from that file

```ts
// hello.ts
export type Greeting = string;

export const hello: Greeting = "Hello World";

// this is a side-effect
window.hello = hello;
```

```ts
import type { Greeting } from './hello'; // only get type

const hello: Greeting = "Hello Mark";
console.log(hello); // "Hello Mark"
console.log(window.hello); // undefined
```

```ts
import { Greeting, hello } from './hello'; // get type and value

console.log(hello); // "Hello World"
console.log(window.hello); // "Hello World"
```

```ts
import type { hello } from './hello'; 

console.log(hello); // Error 'hello' cannot be used as a value because it was imported using 'import type'
```

So in short: Especially importing between config/ui/state I try to default to `import type` in order to not accidentally violate the architectural separation of values

**Comment by mlevison** [comment]
> These are the main benefits we gain:
> 
> * Game state can be executed without needing react or the UI (headless)
> * Configs can be changed easily ([#103 (comment)](https://github.com/mlevison/high-performance-teams-game/issues/103#issuecomment-790055608)) (This will also make the test setup much easier, you'll see)

I can see that
>
> The `GremlinId` and `ActionIds` were declared this way to enable type-checking. Since the state machine can now work with different configs, it can not have an internal understanding of the string literals otherwise it would be coupled again to the config.
> I noted that the UI and state never used these IDs as literals (there is no `dispatch({ type: 'SELECT_ACTION', payload: 'SOME_ACTION_ID' })`)
> The intellisense was only used in tests (and potential future programatic used) where we bypass the UI to select a specific action or gremlin.

Yes and that's were I spend most of my time. They were also used in Gremlins and anywhere else where was in interaction with actions.

> That said, I have Ideas how to revive the types in these scenarios, let's see if they work out.
I look to seeing what works.

> > I mourn the passing of gameConstants - This change feels very odd. Anychance we can create a new version?
> 
> Actually I'm really happy they're gone. This was the prior way of "configuring" the app and a selection of stuff I had not found a better place for :).
> All the infos are now derived from the config and I think that's elegant.
> 
> ```ts
> export const START_CAPACITY = 10; // moved to config.round[0].effect
> export const START_USER_STORY_CHANCE = 30; // moved to config.round[0].#effect
> export const START_GREMLIN_CHANCE = 0; // moved to config.round[0].#effect
> export const GAME_ROUNDS = 8; // derived from config.rounds.length
> export const TRAILING_ROUNDS = 6;  // moved to config.trailingRounds
> export const EFFECT_HIDDEN = Symbol('EFFECT_HIDDEN');  // removed 
> 
> export const TOTAL_ROUNDS = GAME_ROUNDS + TRAILING_ROUNDS; // derived from config.rounds.length + config.trailingRounds
> ```

Some thoughts on constants:
- Never name a file constants - by default some other file gets imported instead.
- A benefit that was lost -> these were named and explicit, making it easier for the reader to understand
- The test helper functions now rely on magic numbers instead of explicit constants. So if we tweak the numbers as a we learn from the MonteCarlo Sim - all the tests will break

I mostly followed the explanation on import types.

**Comment by Xiphe** [comment]
> Context: Lost ID-Literals
> Yes and that's were I spend most of my time. They were also used in Gremlins and anywhere else where was in interaction with actions.

Good point, I have not thought on the internal dependencies. Will try to make this work again

> Never name a file constants - by default some other file gets imported instead.

Agree

> - A benefit that was lost -> these were named and explicit, making it easier for the reader to understand
> - The test helper functions now rely on magic numbers instead of explicit constants. So if we tweak the numbers as a we learn from the MonteCarlo Sim - all the tests will break

I'll refactor the test-helper functions to also work on a config instead of using implicit knowledge about the game
